### ISSUE TITLE
NoSuchMethodError on package block and AnyVal

### ISSUE THREAD

**Issue by ocadaruma** [scala/bug#10783]
- Scala version: 2.12.3
- Steps to reproduce:

```scala
// save as Test.scala
package com.example {
  object X {
    def bar: Int = (new Value(42)).foo
  }
}

package com.example {
  class Value(val value: Int) extends AnyVal {
    def foo: Int = value + 1
  }
}

object Main {
  import com.example._

  def main(args: Array[String]): Unit = {
    println(X.bar)
  }
}
```

```bash
$ scalac Test.scala
$ scala Main
java.lang.NoSuchMethodError: com.example.Value$.foo$extension(I)I
	at com.example.X$.bar(Test.scala:3)
	at Main$.main(Test.scala:17)
	at Main.main(Test.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at scala.reflect.internal.util.ScalaClassLoader.$anonfun$run$2(ScalaClassLoader.scala:98)
	at scala.reflect.internal.util.ScalaClassLoader.asContext(ScalaClassLoader.scala:32)
	at scala.reflect.internal.util.ScalaClassLoader.asContext$(ScalaClassLoader.scala:30)
	at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:129)
	at scala.reflect.internal.util.ScalaClassLoader.run(ScalaClassLoader.scala:98)
	at scala.reflect.internal.util.ScalaClassLoader.run$(ScalaClassLoader.scala:90)
	at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:129)
	at scala.tools.nsc.CommonRunner.run(ObjectRunner.scala:22)
	at scala.tools.nsc.CommonRunner.run$(ObjectRunner.scala:21)
	at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:39)
	at scala.tools.nsc.CommonRunner.runAndCatch(ObjectRunner.scala:29)
	at scala.tools.nsc.CommonRunner.runAndCatch$(ObjectRunner.scala:28)
	at scala.tools.nsc.ObjectRunner$.runAndCatch(ObjectRunner.scala:39)
	at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:66)
	at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:85)
	at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:96)
	at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:101)
	at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
```

- This issue seems to be related to #9922 .
- If I remove `extends AnyVal`, the issue disappears.
- If I swap the order of two packaage blocks, the issue disappears too.

**Comment by som-snytt** [comment]
```
scala> :pa -raw
// Entering paste mode (ctrl-D to finish)

package com.example {
  object X {
    def bar: Int = (new Value(42)).foo
  }
}

package com.example {
  class Value(val value: Int) extends AnyVal {
    def foo: Int = value + 1
  }
}

object Main {
  import com.example._

  def main(args: Array[String]): Unit = {
    println(X.bar)
  }
}

// Exiting paste mode, now interpreting.


scala> Main main null
java.lang.NoSuchMethodError: com.example.Value$.foo$extension(I)I
  at com.example.X$.bar(<pastie>:3)
  at Main$.main(<pastie>:17)
  ... 28 elided

scala> 
```

**Comment by hrhino** [comment]
This probably deserves a similar fix to #6734. The synthetic companion object is getting put in the wrong package block, and extmethods phase isn't finding where to stick the extmethods themselves.

```scala
scala> :pa -raw
// Entering paste mode (ctrl-D to finish)

package com.example {
  object X {
    def bar: Int = (new Value(42)).foo
  }
}

package com.example {
  class Value(val value: Int) extends AnyVal {
    def foo: Int = value + 1
  }
}

object Main {
  import com.example._

  def main(args: Array[String]): Unit = {
    println(X.bar)
  }
}


// Exiting paste mode, now interpreting.

[[syntax trees at end of                     typer]] // <pastie>
package <empty> {
  package com.example {
    object X extends scala.AnyRef {
      def <init>(): com.example.X.type = {
        X.super.<init>();
        ()
      };
      def bar: Int = new Value(42).foo
    };
    <synthetic> object Value extends AnyRef {
      def <init>(): com.example.Value.type = {
        Value.super.<init>();
        ()
      }
    }
  };
  package com.example {
    final class Value extends scala.AnyVal {
      <paramaccessor> private[this] val value: Int = _;
      <stable> <accessor> <paramaccessor> def value: Int = Value.this.value;
      def <init>(value: Int): com.example.Value = {
        Value.super.<init>();
        ()
      };
      def foo: Int = Value.this.value.+(1);
      override <synthetic> def hashCode(): Int = Value.this.value.hashCode();
      override <synthetic> def equals(x$1: Any): Boolean = x$1 match {
  case (_: com.example.Value) => true
  case _ => false
}.&&({
        <synthetic> val Value$1: com.example.Value = x$1.asInstanceOf[com.example.Value];
        Value.this.value.==(Value$1.value)
      })
    }
  };
  object Main extends scala.AnyRef {
    def <init>(): Main.type = {
      Main.super.<init>();
      ()
    };
    import com.example._;
    def main(args: Array[String]): Unit = scala.Predef.println(com.example.X.bar)
  }
}

[[syntax trees at end of                extmethods]] // <pastie>
package <empty> {
  package com.example {
    object X extends scala.AnyRef {
      def <init>(): com.example.X.type = {
        X.super.<init>();
        ()
      };
      def bar: Int = new Value(42).foo
    };
    <synthetic> object Value extends AnyRef {
      def <init>(): com.example.Value.type = {
        Value.super.<init>();
        ()
      }
    }
  };
  package com.example {
    final class Value extends scala.AnyVal {
      <paramaccessor> private[this] val value: Int = _;
      <stable> <accessor> <paramaccessor> def value: Int = Value.this.value;
      def <init>(value: Int): com.example.Value = {
        Value.super.<init>();
        ()
      };
      def foo: Int = example.this.Value.foo$extension(Value.this);
      override <synthetic> def hashCode(): Int = example.this.Value.hashCode$extension(Value.this)();
      override <synthetic> def equals(x$1: Any): Boolean = example.this.Value.equals$extension(Value.this)(x$1)
    }
  };
  object Main extends scala.AnyRef {
    def <init>(): Main.type = {
      Main.super.<init>();
      ()
    };
    import com.example._;
    def main(args: Array[String]): Unit = scala.Predef.println(com.example.X.bar)
  }
}
```

**Comment by dwijnand** [comment]
Fixed in https://github.com/scala/scala/pull/6443.
